### PR TITLE
GcsToSourceDB template integration test - Spanner Reverse replication writer IT

### PR DESCRIPTION
**Pull request by darshan-sj**

### CODE REVIEW

File: v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbIT.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+    spannerResourceManager = createSpannerDatabase();
+    spannerMetadataResourceManager = createSpannerMetadataDatabase();
+
+    // Run a simple IT
+    simpleGcsToSourceDbTest(
+        testName,
+        TABLE,
+        mySQLSource,
+        DatastreamResourceManager.DestinationOutputFormat.AVRO_FILE_FORMAT,
```
Comment (aksharauke) [inline v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbIT.java:127]: Did not understand why we need Datastream resource

File: v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbIT.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            put("GCSInputDirectoryPath", getGcsPath("cdc"));
+          }
+        };
+
+    // Construct template
+    String jobName = PipelineUtils.createJobName(testName);
+    // /-DunifiedWorker=true when using runner v2
+    LaunchConfig.Builder options = LaunchConfig.builder(jobName, specPath);
+    options.setParameters(params);
+    options.addEnvironment("additionalExperiments", Collections.singletonList("use_runner_v2"));
```
Comment (aksharauke) [inline v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbIT.java:214]: The writer template does not run with v2

File: v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbIT.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    assertThat(rowCount).isEqualTo(1);
+    List<Map<String, Object>> rows = jdbcResourceManager.readTable(TABLE);
+    assertThat(rows).hasSize(1);
+    assertThat(rows.get(0).getOrDefault("id", -1)).isEqualTo(1);
+    assertThat(rows.get(0).getOrDefault("name", "non existent")).isEqualTo("FF");
+  }
+
+  private void launchWriterDataflowJob() throws IOException {
+    // default parameters
+    Map<String, String> params =
```
Comment (aksharauke) [inline v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbIT.java:194]: If the reader template is not there, the writer template needs to know the start time and the window duration. It needs that to check for the GCS files. The GCSReader class has the logic to construct the file name.

File: v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbIT.java [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+      Function<LaunchConfig.Builder, LaunchConfig.Builder> paramsAdder)
+      throws IOException, InterruptedException {
+
+    createAndUploadShardConfigToGcs();
+    gcsClient.uploadArtifact(
+        "input/session.json", Resources.getResource(SESSION_FILE_RESOURSE).getPath());
+    launchWriterDataflowJob();
+
+    // Write events to GCS
+    gcsClient.uploadArtifact(
```
Comment (aksharauke) [inline v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbIT.java:175]: The file name is expected to be in a specific naming format. Please refer the GCSReader.

**Review by aksharauke — APPROVED** [review]

File: v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbWithoutReaderIT.java [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+   *
+   * @throws IOException
+   */
+  @Before
+  public void setUp() throws IOException {
+    skipBaseCleanup = true;
+    synchronized (GCSToSourceDbWithoutReaderIT.class) {
+      testInstances.add(this);
+      if (jobInfo == null) {
+        spannerResourceManager = createSpannerDatabase(SPANNER_DDL_RESOURCE);
```
Comment (aksharauke) [inline v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbWithoutReaderIT.java:84]: For writer, this is not needed

File: v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbWithoutReaderIT.java [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    shard.setDbName(jdbcResourceManager.getDatabaseName());
+    JsonObject jsObj = (JsonObject) new Gson().toJsonTree(shard).getAsJsonObject();
+    jsObj.remove("secretManagerUri"); // remove field secretManagerUri
+    JsonArray ja = new JsonArray();
+    ja.add(jsObj);
+    String shardFileContents = ja.toString();
+    LOG.info("Shard file contents: {}", shardFileContents);
+    gcsResourceManager.createArtifact("input/shard.json", shardFileContents);
+  }
+}
```
Comment (manitgupta) [inline v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbWithoutReaderIT.java:225]: Are there more expected ITs for `GCSToSourcdDB*`? If yes, we should put this in a `TestBase` class similar to our approach in `DataStreamToSpanner`

File: v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbWithoutReaderIT.java (hunk 5 above)
Comment (darshan-sj) [reply v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbWithoutReaderIT.java]: Done. Removed the main spanner resource manager

File: v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbWithoutReaderIT.java (hunk 6 above)
Comment (darshan-sj) [reply v2/gcs-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/GCSToSourceDbWithoutReaderIT.java]: Yes. I'll move out the common methods in next PRs along with the actual test cases.

**Pull request merged**